bundles/org.openhab.binding.allplay/src/main/java/org/openhab/binding/allplay/internal/AllPlayAudioSink.java

```diff
@@ -75,12 +71,36 @@ public String getLabel(Locale locale) {
     }
 
     @Override
-    public void process(AudioStream audioStream)
+    protected void processAsynchronously(@Nullable AudioStream audioStream)
             throws UnsupportedAudioFormatException, UnsupportedAudioStreamException {
+        if (audioStream == null) {
+            return;
+        }
+        String url;
+        if (audioStream instanceof URLAudioStream urlAudioStream) {
+            // it is an external URL, the speaker can access it itself and play it
+            url = urlAudioStream.getURL();
+            tryClose(audioStream);
+        } else if (callbackUrl != null) {
+            StreamServed streamServed;
+            try {
+                streamServed = audioHTTPServer.serve(audioStream, 10, true);
+            } catch (IOException e) {
+                tryClose(audioStream);
+                throw new UnsupportedAudioStreamException(
+                        "AllPlay was not able to handle the audio stream (cache on disk failed).",
+                        audioStream.getClass(), e);
+            }
+            url = callbackUrl + streamServed.url();
+            streamServed.playEnd().thenRun(() -> this.playbackFinished(audioStream));
+        } else {
+            logger.warn("We do not have any callback url, so AllPlay cannot play the audio stream!");
+            tryClose(audioStream);
+            return;
+        }
         try {
-            String url = convertAudioStreamToUrl(audioStream);
             handler.playUrl(url);
-        } catch (SpeakerException | AllPlayCallbackException e) {
+        } catch (SpeakerException e) {
```

Comment:
Line below, consider checking for debug log level and provide only full stack trace in this case, otherwise `e.getMessage()`.

Reply:
Done

Reply:
Ok, changed.